src/client/dfuse/ops/read.c

```diff
@@ -24,31 +24,129 @@
 #include "dfuse_common.h"
 #include "dfuse.h"
 
+#define READAHEAD_SIZE (1024 * 1024)
+
 void
 dfuse_cb_read(fuse_req_t req, fuse_ino_t ino, size_t len, off_t position,
 	      struct fuse_file_info *fi)
 {
 	struct dfuse_obj_hdl		*oh = (struct dfuse_obj_hdl *)fi->fh;
-	d_iov_t				iov = {};
+	struct dfuse_projection_info	*fs_handle = fuse_req_userdata(req);
+	const struct fuse_ctx		*fc = fuse_req_ctx(req);
+	d_iov_t				iov[2] = {};
 	d_sg_list_t			sgl = {};
+	struct fuse_bufvec		fb = {};
 	daos_size_t			size;
 	void				*buff;
+	void				*ahead_buff = NULL;
 	int				rc;
 
+	DFUSE_TRA_INFO(oh, "%#zx-%#zx requested pid=%d",
+		       position, position + len - 1, fc->pid);
+
+	DFUSE_TRA_DEBUG(oh, "Will try readahead file size %zi",
+			oh->doh_ie->ie_stat.st_size);
+
 	D_ALLOC(buff, len);
 	if (!buff) {
 		DFUSE_REPLY_ERR_RAW(NULL, req, ENOMEM);
 		return;
 	}
 
+	/* TODO:
+	 *
+	 * fuse_lowlevel_notify_store() does not call free() on the buffer
+	 * after use so a single buffer rather than two would work well here.
+	 * We could combine the truncated and readahead logic by just having
+	 * the trunctated check set a "skip-dfs_read" flag and just have one
+	 * flow through this function rather than two.
+	 * For the ie_truncated code a lot of zeroed data is being passed to
+	 * the kernel so it would be possible to have a "zero-buffer" allcocated
+	 * once and reuse it rather than a malloc + memset every time.
+	 */
+
```

Comment:
We should probably rework this part a little bit. There should be ways to simplify it.

Reply:
Yes, I already re-wrote it once because I thought it needed two allocation points.  I was trying to get to a point where it was stable and reflected what you'd been testing first before changing it again.